encoder/libav_encoder.cpp

```diff
@@ -0,0 +1,548 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (C) 2022, Raspberry Pi Ltd
+ *
+ * libav_encoder.cpp - libav video encoder.
+ */
+
+#include <fcntl.h>
+#include <poll.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+
+#include <libdrm/drm_fourcc.h>
+#include <linux/videodev2.h>
+
+#include <chrono>
+#include <iostream>
+
+#include "libav_encoder.hpp"
+
+void LibAvEncoder::initVideoCodec(VideoOptions const *options, StreamInfo const &info)
+{
+	const std::string codec_name("h264_v4l2m2m");
+
+	const AVCodec *codec = avcodec_find_encoder_by_name(codec_name.c_str());
+	if (!codec)
+		throw std::runtime_error("libav: cannot find video encoder " + codec_name);
+
+	codec_ctx_[Video] = avcodec_alloc_context3(codec);
+	if (!codec_ctx_[Video])
+		throw std::runtime_error("libav: Cannot allocate video context");
+
+	codec_ctx_[Video]->width = info.width;
+	codec_ctx_[Video]->height = info.height;
+	// usec timebase
+	codec_ctx_[Video]->time_base = { 1, 1000 * 1000 };
+	codec_ctx_[Video]->framerate = { (int)(options->framerate * 1000), 1000 };
```

Comment:
I assume these are just rational numbers. I wonder if there's some syntactic sugar that would make this more obvious in some way, but I'm probably just fussing because, frankly, most of this is going over my head!!

Reply:
Yes, they are rational numbers.  Essentially, I want to match the units of the frame timestamps for simplicity.